sys/vm/vm_kern.c

```diff
@@ -350,6 +356,7 @@ kmem_alloc_contig_domain(int domain, vm_size_t size, int flags, vm_paddr_t low,
 	vmem = vm_dom[domain].vmd_kernel_arena;
 	if (vmem_alloc(vmem, asize, flags | M_BESTFIT, &addr))
 		return (NULL);
+	addr = cheri_kern_andperm(addr, CHERI_PERMS_KERNEL_DATA);
```

Comment:
It is weird and probably incorrect that

1) kmem_alloc_attr handles M_EXEC but uses the same vmem arena regardless,
2) kmem_malloc handles M_EXEC and correctly uses a separate vmem arena for that purpose,
3) kmem_alloc_contig doesn't handle M_EXEC,

but that's an upstream problem. Though, I think it might be nice if the vmem arena could automatically limit permissions before returning from vmem_alloc().

Reply:
I'm not sure which would be worse/better for vmem_alloc().  Possibly though if M_EXEC always uses the separate vmem arena we could make the vmem_alloc() case work.

Reply:
Yeah, I think kmem_* should always be using the separate arena for executable mappings and I need to look at why it doesn't. Anyway, that's a separate issue.